File: src/main/java/com/example/librarymanagement/dto/mapper/BookPresenceMapper.java
package com.example.librarymanagement.dto.mapper;

import com.example.librarymanagement.dto.BookPresenceDto;
import com.example.librarymanagement.model.entity.BookPresence;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

@Component
public class BookPresenceMapper {
    private final UserMapper userMapper;

    public BookPresenceMapper(UserMapper userMapper) {
        this.userMapper = userMapper;
    }

    public BookPresenceDto toBookPresenceDto(BookPresence bookPresence) {
        BookPresenceDto bookPresenceDto = new BookPresenceDto();
        bookPresenceDto.setId(bookPresence.getId());
        bookPresenceDto.setBookTitle(bookPresence.getBook().getTitle());
        bookPresenceDto.setBookAuthorId(bookPresence.getBook().getAuthor().getId());
        bookPresenceDto.setLibraryNameId(bookPresence.getLibrary().getId());
        bookPresenceDto.setUser(userMapper.toUserResponseDto(bookPresence.getUser()));
        bookPresenceDto.setAddress(bookPresence.getLibrary().getAddress());
        bookPresenceDto.setAvailability(bookPresence.getAvailability());
        return bookPresenceDto;
    }

    public List<BookPresenceDto> toBookPresenceDto(List<BookPresence> bookPresenceList) {
        if(bookPresenceList == null || bookPresenceList.isEmpty()) return new ArrayList<>();

        return bookPresenceList.stream()
                .map(this::toBookPresenceDto)
                .toList();
    }
}

File: src/main/java/com/example/librarymanagement/dto/mapper/BookPresenceMapper.kt
package com.example.librarymanagement.dto.mapper

import com.example.librarymanagement.dto.BookPresenceDto
import com.example.librarymanagement.model.domain.BookPresence
import org.springframework.stereotype.Component

@Component
class BookPresenceMapper {
    fun toBookPresenceDto(bookPresence: BookPresence): BookPresenceDto {
        return BookPresenceDto(
            id = bookPresence.id,
            bookId = bookPresence.bookId,
            libraryId = bookPresence.libraryId,
            userId = bookPresence.userId,
            availability = bookPresence.availability
        )
    }

    fun toBookPresenceDto(bookPresenceList: List<BookPresence?>): List<BookPresenceDto> {
        return bookPresenceList.map { toBookPresenceDto(it!!) }
    }
}
